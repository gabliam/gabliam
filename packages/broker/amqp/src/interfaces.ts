import { ConsumeMessage, Options } from 'amqplib';

export interface QueueOptions {
  /**
   * if true, scopes the queue to the connection (defaults to false)
   */
  exclusive: boolean;

  /**
   * if true, the queue will survive broker restarts,
   * modulo the effects of exclusive and autoDelete;
   * this defaults to true if not supplied, unlike the others
   */
  durable: boolean;

  /**
   * if true, the queue will be deleted when the number of consumers drops to zero (defaults to false)
   */
  autoDelete: boolean;

  /**
   * RabbitMQ extensions can also be supplied as options.
   * These typically require non-standard x-* keys and values, sent in the arguments table; e.g., 'x-expires'.
   * When supplied in options, the x- prefix for the key is removed; e.g., 'expires'.
   * Values supplied in options will overwrite any analogous field you put in options.arguments.
   */
  arguments?: { [name: string]: string };
}

export interface QueueConfiguration {
  queueName: string;

  options: QueueOptions;
}

export type ConsumerHandler = (
  msg: ConsumeMessage | null
) => void | Promise<void>;

export type Handler = (...args: any[]) => any | Promise<any>;

export interface Controller {
  [k: string]: Handler;
}

export type ConsumeOptions = Options.Consume;

export type SendOptions = Options.Publish;

export interface RabbitHandlerOptions {
  consumeOptions?: ConsumeOptions;

  sendOptions?: SendOptions;

  sendOptionsError?: SendOptions;
}

export interface ConsumeConfig {
  queueName: string;
  handler: ConsumerHandler;
  options?: ConsumeOptions;
}

export interface QueueDictionnary {
  [k: string]: QueueConfiguration;
}

/**
 * Configuration for a connection
 */
export interface ConnectionConfig {
  /**
   * Name of the connection
   */
  name: string;

  /**
   * Url of the connection
   */
  url: string;

  /**
   * Value passed to rabbitmq for an undefined value
   * amqplib can not send an undefined value because the lib uses Buffer
   *
   */
  undefinedValue: string;

  queues: QueueDictionnary;

  gzipEnabled: boolean;
}
