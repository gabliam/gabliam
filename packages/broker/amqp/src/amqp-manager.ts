import { ConnectionConfig, QueueDictionnary } from './interfaces';
import { AmqpConnection } from './amqp-connection';
import { Queue } from './queue';
import { ValueExtractor } from '@gabliam/core';
import {
  AmqpDuplicateConnectionError,
  AmqpConnectionNotFoundError,
} from './errors';

export class AmqpConnectionManager {
  private connections: AmqpConnection[] = [];

  constructor(
    connectionConfigs: ConnectionConfig[],
    valueExtractor: ValueExtractor
  ) {
    for (const [
      index,
      { name, url, queues, undefinedValue, gzipEnabled },
    ] of connectionConfigs.entries()) {
      if (this.connections.find(c => c.name === name) !== undefined) {
        throw new AmqpDuplicateConnectionError(name);
      }

      this.connections.push(
        new AmqpConnection(
          index,
          name,
          url,
          undefinedValue,
          this.createQueue(queues),
          valueExtractor,
          gzipEnabled
        )
      );
    }
  }

  async start() {
    await Promise.all(this.connections.map(c => c.start()));
  }

  async stop() {
    await Promise.all(this.connections.map(c => c.stop()));
  }

  getConnection(name: string) {
    const connection = this.connections.find(c => c.name === name);
    if (!connection) {
      throw new AmqpConnectionNotFoundError(name);
    }
    return connection;
  }

  getDefaultConnection() {
    const connection = this.connections.find(c => c.name === 'default');
    if (connection === undefined) {
      return this.connections[0];
    } else {
      return connection;
    }
  }

  private createQueue(queueConfig: QueueDictionnary = {}) {
    return Object.keys(queueConfig).map<Queue>((k: string) => {
      const q = queueConfig[k];
      return new Queue(q.queueName, q.options);
    });
  }
}
