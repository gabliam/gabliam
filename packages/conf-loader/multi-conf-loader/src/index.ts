import {
  configResolver,
  LoaderConfigParseError,
  Resolver,
} from '@gabliam/core';
import d from 'debug';
import fs from 'fs';
import yaml from 'js-yaml';
import _ from 'lodash';
import { inspect, promisify } from 'util';

// Promisify
const glob = promisify(require('glob')); // require and no import for typings bug
const readFile = promisify(fs.readFile);

const debug = d('@gabliam/multi-conf-loader');

async function loadConstant(
  folder: string,
  projectName: string,
  resolver: Resolver,
  profile?: string,
) {
  const files: string[] = await glob(`**/constants?(-+([a-zA-Z])).@(yml)`, {
    cwd: folder,
  });

  let constants: object = {};

  if (!files || files.length === 0) {
    return constants;
  }

  const defaultProfileFile = files.find((file) => file === `constants.yml`);
  if (defaultProfileFile) {
    constants = _.merge(
      {},
      constants,
      await loadFile(`${folder}/${defaultProfileFile}`, resolver),
    );
  }

  if (profile) {
    const profileFile = files.find(
      (file) => file === `constants-${profile}.yml`,
    );

    if (profileFile) {
      constants = _.merge(
        {},
        constants,
        await loadFile(`${folder}/${profileFile}`, resolver),
      );
    }
  }
  return constants;
}

/**
 * FileLoader
 */
const multiConfLoader = async (
  { folder, projectName }: { folder: string; projectName: string },
  profile?: string,
) => {
  debug('loadConfig', folder);
  const resolver = configResolver(folder);
  const constants = await loadConstant(folder, projectName, resolver, profile);

  // create tag config
  const ConfigYamlType = new yaml.Type('!config', {
    kind: 'scalar',
    construct: function (data) {
      return _.get(constants, data, {});
    },
    instanceOf: Object,
  });

  const GAB_SCHEMA = yaml.Schema.create([ConfigYamlType]);

  const files: string[] = await glob(`**/application?(-+([a-zA-Z])).@(yml)`, {
    cwd: folder,
  });

  let config: object = {};

  if (!files || files.length === 0) {
    return config;
  }

  const defaultProfileFile = files.find((file) => file === `application.yml`);
  if (defaultProfileFile) {
    config = _.merge(
      {},
      config,
      await loadConfigProject(
        `${folder}/${defaultProfileFile}`,
        projectName,
        GAB_SCHEMA,
        resolver,
      ),
    );
  }

  if (profile) {
    const profileFile = files.find(
      (file) => file === `application-${profile}.yml`,
    );

    if (profileFile) {
      config = _.merge(
        {},
        config,
        await loadConfigProject(
          `${folder}/${profileFile}`,
          projectName,
          GAB_SCHEMA,
          resolver,
        ),
      );
    }
  }
  debug('loadedConfig', inspect(config, { depth: null }));
  return config;
};

async function loadConfigProject(
  filePath: string,
  projectName: string,
  schema: yaml.Schema,
  resolver: Resolver,
) {
  const data = await readFile(filePath, 'utf8');
  try {
    const vals = yaml.loadAll(data, undefined, { schema }) || [];
    return await resolver(_.find(vals, <any>{ projectName }) || {});
  } catch (e) {
    throw new LoaderConfigParseError(filePath, e);
  }
}

async function loadFile(filePath: string, resolver: Resolver): Promise<Object> {
  const data = await readFile(filePath, 'utf8');
  try {
    return await resolver(yaml.load(data) || {});
  } catch (e) {
    throw new LoaderConfigParseError(filePath, e);
  }
}

export default multiConfLoader;
