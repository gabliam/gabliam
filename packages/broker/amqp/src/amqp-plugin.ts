import {
  Container,
  GabliamPlugin,
  Plugin,
  reflection,
  Registry,
  Scan,
} from '@gabliam/core';
import d from 'debug';
import { AmqpConnectionManager } from './amqp-manager';
import { TYPE } from './constants';
import {
  CUnit,
  RabbitConsumer,
  RabbitHandler,
  RabbitListener,
} from './metadatas';

const debug = d('Gabliam:Plugin:AmqpPlugin');

@Plugin()
@Scan()
export class AmqpPlugin implements GabliamPlugin {
  build(container: Container, registry: Registry) {
    debug('build AmqpPlugin');
    const controllerIds = registry.get(TYPE.RabbitController);
    const connectionManager = container.get(AmqpConnectionManager);
    debug('controllerIds', controllerIds, container);
    controllerIds.forEach(({ id: controllerId }) => {
      const controller = container.get<any>(controllerId);
      debug('controller', controller);

      const handlerMetadatas = reflection.propMetadataOfDecorator<RabbitHandler>(
        controller.constructor,
        RabbitConsumer,
        RabbitListener,
      );

      const [cunit] = reflection
        .annotationsOfDecorator<CUnit>(controller.constructor, CUnit)
        .slice(-1);

      if (Object.keys(handlerMetadatas)) {
        for (const [key, metas] of Object.entries(handlerMetadatas)) {
          const [handlerMetadata] = metas.slice(-1);
          if (cunit) {
            connectionManager
              .getConnection(cunit.name)
              .constructAndAddConsume(key, handlerMetadata, controller);
          } else {
            connectionManager
              .getDefaultConnection()
              .constructAndAddConsume(key, handlerMetadata, controller);
          }
        }
      }
    });
  }

  async start(container: Container) {
    const connection = container.get(AmqpConnectionManager);
    await connection.start();
  }

  async stop(container: Container) {
    const connection = container.get(AmqpConnectionManager);
    await connection.stop();
  }
}
