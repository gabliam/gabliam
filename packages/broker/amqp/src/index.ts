import { AmqpPlugin } from './amqp-plugin';

export * from './amqp-connection';
export * from './amqp-manager';
export * from './errors';
export * from './interfaces';
export * from './metadatas';
export * from './queue';
export { AmqpPlugin as default };
