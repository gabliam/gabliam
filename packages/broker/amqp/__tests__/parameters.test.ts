import { AmqpPluginTest } from './amqp-plugin-test';
import {
  RabbitController,
  RabbitListener,
  RabbitConsumer,
  Message,
  Content,
  Properties,
  Fields,
} from '../src/index';
import { config, configWithGzip } from './conf';
import { AmqpConnection } from '../src/amqp-connection';
import { Gabliam } from '@gabliam/core';
import { Deferred } from './defered';
import * as sinon from 'sinon';

let appTest: AmqpPluginTest;

beforeEach(async () => {
  appTest = new AmqpPluginTest();
});

afterEach(async () => {
  try {
    await appTest.gab.stop();
  } catch (e) {}
  try {
    await appTest.destroy();
  } catch (e) {}
});

[config, configWithGzip].forEach(conf => {
  const describeName = conf.gzipEnabled ? 'Parameters gzip:' : 'Parameters:';
  describe(describeName, () => {
    test('should bind a method parameter to message', async () => {
      const d = new Deferred();

      @RabbitController()
      class ControllerTest {
        @RabbitConsumer('consumerTest')
        consumer(@Message() msg: any) {
          return msg;
        }

        @RabbitListener('listenerTest')
        listener(@Message() msg: any) {
          d.resolve(msg);
        }
      }
      const spyConsumer = sinon.spy(ControllerTest.prototype, 'consumer');
      const spyListener = sinon.spy(ControllerTest.prototype, 'listener');

      appTest.addConf('application.amqp', conf);
      appTest.addClass(ControllerTest);
      await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(
        Gabliam
      );
      const connection = appTest.gab.container.get(AmqpConnection);

      const respConsumer = await connection.sendAndReceive(
        'consumerTest',
        'TEST'
      );

      expect(respConsumer).toMatchSnapshot<any>({
        content: expect.any(Object),
        properties: {
          correlationId: expect.any(String),
          replyTo: expect.any(String),
        },
        fields: {
          consumerTag: expect.any(String),
        },
      });
      expect(spyConsumer.callCount).toMatchSnapshot();
      spyConsumer.resetHistory();

      connection.sendToQueue('listenerTest', 'TEST');
      const resp = await d.promise;
      expect(resp).toMatchSnapshot<any>({
        content: expect.any(Object),
        fields: {
          consumerTag: expect.any(String),
        },
      });
      expect(spyListener.callCount).toMatchSnapshot();
      spyListener.resetHistory();
    });

    [
      {
        name: 'to the content object and call with buffer string',
        content: new Buffer('buffertest'),
        args: undefined,
      },
      {
        name: 'to the content key not exist and call with buffer string',
        content: new Buffer('buffertest'),
        args: 'test',
      },
      {
        name: 'to the content object and with buffer object',
        content: new Buffer(JSON.stringify({ test: 'buffertest' })),
        args: undefined,
      },
      {
        name: 'to the content key not exist and with buffer object',
        content: new Buffer(JSON.stringify({ test: 'buffertest' })),
        args: 'testlol',
      },
      {
        name: 'to the content object and with string',
        content: 'test',
        args: undefined,
      },
      {
        name: 'to the content key not exist and with string',
        content: 'test',
        args: 'test',
      },
      {
        name: 'to the content object and with object',
        content: { test: 'cool' },
        args: undefined,
      },
      {
        name: 'to the content and with object',
        content: { test: 'cool' },
        args: 'test',
      },
      {
        name: 'to the content key not exist and with object',
        content: { test: 'cool' },
        args: 'testlol',
      },
    ].forEach(testCase => {
      test(`should bind a method parameter ${testCase.name}`, async () => {
        const d = new Deferred();

        @RabbitController()
        class ControllerTest {
          @RabbitConsumer('consumerTest')
          consumer(@Content(testCase.args) content: any) {
            return content;
          }

          @RabbitListener('listenerTest')
          listener(@Content(testCase.args) content: any) {
            d.resolve(content);
          }
        }
        const spyConsumer = sinon.spy(ControllerTest.prototype, 'consumer');
        const spyListener = sinon.spy(ControllerTest.prototype, 'listener');

        appTest.addConf('application.amqp', conf);
        appTest.addClass(ControllerTest);
        await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(
          Gabliam
        );
        const connection = appTest.gab.container.get(AmqpConnection);

        const respConsumer = await connection.sendAndReceive(
          'consumerTest',
          testCase.content
        );
        expect(respConsumer).toMatchSnapshot();
        expect(spyConsumer.callCount).toMatchSnapshot();
        spyConsumer.resetHistory();

        connection.sendToQueue('listenerTest', testCase.content);
        const resp = await d.promise;
        expect(resp).toMatchSnapshot();
        expect(spyListener.callCount).toMatchSnapshot();
        spyListener.resetHistory();
      });
    });

    test(`should bind a method to Properties object`, async () => {
      const d = new Deferred();

      @RabbitController()
      class ControllerTest {
        @RabbitConsumer('consumerTest')
        consumer(@Properties() content: any) {
          return content;
        }

        @RabbitListener('listenerTest')
        listener(@Properties() content: any) {
          d.resolve(content);
        }
      }
      const spyConsumer = sinon.spy(ControllerTest.prototype, 'consumer');
      const spyListener = sinon.spy(ControllerTest.prototype, 'listener');

      appTest.addConf('application.amqp', conf);
      appTest.addClass(ControllerTest);
      await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(
        Gabliam
      );
      const connection = appTest.gab.container.get(AmqpConnection);

      const respConsumer = await connection.sendAndReceive(
        'consumerTest',
        'test'
      );
      expect(respConsumer).toMatchSnapshot({
        correlationId: expect.any(String),
        replyTo: expect.any(String),
      });
      expect(spyConsumer.callCount).toMatchSnapshot();
      spyConsumer.resetHistory();

      connection.sendToQueue('listenerTest', 'test');
      const resp = await d.promise;
      expect(resp).toMatchSnapshot();
      expect(spyListener.callCount).toMatchSnapshot();
      spyListener.resetHistory();
    });

    test(`should bind a method to Properties headers `, async () => {
      const d = new Deferred();

      @RabbitController()
      class ControllerTest {
        @RabbitConsumer('consumerTest')
        consumer(@Properties('headers') content: any) {
          return content;
        }

        @RabbitListener('listenerTest')
        listener(@Properties('headers') content: any) {
          d.resolve(content);
        }
      }
      const spyConsumer = sinon.spy(ControllerTest.prototype, 'consumer');
      const spyListener = sinon.spy(ControllerTest.prototype, 'listener');

      appTest.addConf('application.amqp', conf);
      appTest.addClass(ControllerTest);
      await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(
        Gabliam
      );
      const connection = appTest.gab.container.get(AmqpConnection);

      const respConsumer = await connection.sendAndReceive(
        'consumerTest',
        'test',
        {
          headers: {
            'api-auth': 'aze&ée((("a&é"&é"',
          },
        }
      );
      expect(respConsumer).toMatchSnapshot();
      expect(spyConsumer.callCount).toMatchSnapshot();
      spyConsumer.resetHistory();

      connection.sendToQueue('listenerTest', 'test', {
        headers: {
          'api-auth': 'aze&ée((("a&é"&é"',
        },
      });
      const resp = await d.promise;
      expect(resp).toMatchSnapshot();
      expect(spyListener.callCount).toMatchSnapshot();
      spyListener.resetHistory();
    });

    test(`should bind a method to Fields object`, async () => {
      const d = new Deferred();

      @RabbitController()
      class ControllerTest {
        @RabbitConsumer('consumerTest')
        consumer(@Fields() content: any) {
          return content;
        }

        @RabbitListener('listenerTest')
        listener(@Fields() content: any) {
          d.resolve(content);
        }
      }
      const spyConsumer = sinon.spy(ControllerTest.prototype, 'consumer');
      const spyListener = sinon.spy(ControllerTest.prototype, 'listener');

      appTest.addConf('application.amqp', conf);
      appTest.addClass(ControllerTest);
      await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(
        Gabliam
      );
      const connection = appTest.gab.container.get(AmqpConnection);

      const respConsumer = await connection.sendAndReceive(
        'consumerTest',
        'test'
      );
      expect(respConsumer).toMatchSnapshot({
        consumerTag: expect.any(String),
        deliveryTag: expect.any(Number),
        exchange: expect.any(String),
        redelivered: expect.any(Boolean),
      });
      expect(spyConsumer.callCount).toMatchSnapshot();
      spyConsumer.resetHistory();

      connection.sendToQueue('listenerTest', 'test');
      const resp = await d.promise;
      expect(resp).toMatchSnapshot({
        consumerTag: expect.any(String),
        deliveryTag: expect.any(Number),
        exchange: expect.any(String),
        redelivered: expect.any(Boolean),
      });
      expect(spyListener.callCount).toMatchSnapshot();
      spyListener.resetHistory();
    });

    test(`should bind a method to Fields routingKey `, async () => {
      const d = new Deferred();

      @RabbitController()
      class ControllerTest {
        @RabbitConsumer('consumerTest')
        consumer(@Fields('routingKey') content: any) {
          return content;
        }

        @RabbitListener('listenerTest')
        listener(@Fields('routingKey') content: any) {
          d.resolve(content);
        }
      }
      const spyConsumer = sinon.spy(ControllerTest.prototype, 'consumer');
      const spyListener = sinon.spy(ControllerTest.prototype, 'listener');

      appTest.addConf('application.amqp', conf);
      appTest.addClass(ControllerTest);
      await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(
        Gabliam
      );
      const connection = appTest.gab.container.get(AmqpConnection);

      const respConsumer = await connection.sendAndReceive(
        'consumerTest',
        'test'
      );
      expect(respConsumer).toMatchSnapshot();
      expect(spyConsumer.callCount).toMatchSnapshot();
      spyConsumer.resetHistory();

      connection.sendToQueue('listenerTest', 'test');
      const resp = await d.promise;
      expect(resp).toMatchSnapshot();
      expect(spyListener.callCount).toMatchSnapshot();
      spyListener.resetHistory();
    });
  });
});
