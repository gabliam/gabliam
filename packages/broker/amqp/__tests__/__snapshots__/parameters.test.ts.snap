// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parameters gzip: should bind a method parameter to message 1`] = `
Object {
  "content": Any<Object>,
  "fields": Object {
    "consumerTag": Any<String>,
    "deliveryTag": 1,
    "exchange": "",
    "redelivered": false,
    "routingKey": "consumerTest",
  },
  "properties": Object {
    "contentEncoding": "gzip",
    "contentType": "application/json",
    "correlationId": Any<String>,
    "expiration": "5000",
    "headers": Object {},
    "replyTo": Any<String>,
  },
}
`;

exports[`Parameters gzip: should bind a method parameter to message 2`] = `1`;

exports[`Parameters gzip: should bind a method parameter to message 3`] = `
Object {
  "content": Any<Object>,
  "fields": Object {
    "consumerTag": Any<String>,
    "deliveryTag": 3,
    "exchange": "",
    "redelivered": false,
    "routingKey": "listenerTest",
  },
  "properties": Object {
    "appId": undefined,
    "clusterId": undefined,
    "contentEncoding": "gzip",
    "contentType": "application/json",
    "correlationId": undefined,
    "deliveryMode": undefined,
    "expiration": undefined,
    "headers": Object {},
    "messageId": undefined,
    "priority": undefined,
    "replyTo": undefined,
    "timestamp": undefined,
    "type": undefined,
    "userId": undefined,
  },
}
`;

exports[`Parameters gzip: should bind a method parameter to message 4`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content and with object 1`] = `"cool"`;

exports[`Parameters gzip: should bind a method parameter to the content and with object 2`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content and with object 3`] = `"cool"`;

exports[`Parameters gzip: should bind a method parameter to the content and with object 4`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and call with buffer string 1`] = `undefined`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and call with buffer string 2`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and call with buffer string 3`] = `undefined`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and call with buffer string 4`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with buffer object 1`] = `undefined`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with buffer object 2`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with buffer object 3`] = `undefined`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with buffer object 4`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with object 1`] = `undefined`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with object 2`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with object 3`] = `undefined`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with object 4`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with string 1`] = `undefined`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with string 2`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with string 3`] = `undefined`;

exports[`Parameters gzip: should bind a method parameter to the content key not exist and with string 4`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content object and call with buffer string 1`] = `"buffertest"`;

exports[`Parameters gzip: should bind a method parameter to the content object and call with buffer string 2`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content object and call with buffer string 3`] = `"buffertest"`;

exports[`Parameters gzip: should bind a method parameter to the content object and call with buffer string 4`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content object and with buffer object 1`] = `
Object {
  "test": "buffertest",
}
`;

exports[`Parameters gzip: should bind a method parameter to the content object and with buffer object 2`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content object and with buffer object 3`] = `
Object {
  "test": "buffertest",
}
`;

exports[`Parameters gzip: should bind a method parameter to the content object and with buffer object 4`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content object and with object 1`] = `
Object {
  "test": "cool",
}
`;

exports[`Parameters gzip: should bind a method parameter to the content object and with object 2`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content object and with object 3`] = `
Object {
  "test": "cool",
}
`;

exports[`Parameters gzip: should bind a method parameter to the content object and with object 4`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content object and with string 1`] = `"test"`;

exports[`Parameters gzip: should bind a method parameter to the content object and with string 2`] = `1`;

exports[`Parameters gzip: should bind a method parameter to the content object and with string 3`] = `"test"`;

exports[`Parameters gzip: should bind a method parameter to the content object and with string 4`] = `1`;

exports[`Parameters gzip: should bind a method to Fields object 1`] = `
Object {
  "consumerTag": Any<String>,
  "deliveryTag": Any<Number>,
  "exchange": Any<String>,
  "redelivered": Any<Boolean>,
  "routingKey": "consumerTest",
}
`;

exports[`Parameters gzip: should bind a method to Fields object 2`] = `1`;

exports[`Parameters gzip: should bind a method to Fields object 3`] = `
Object {
  "consumerTag": Any<String>,
  "deliveryTag": Any<Number>,
  "exchange": Any<String>,
  "redelivered": Any<Boolean>,
  "routingKey": "listenerTest",
}
`;

exports[`Parameters gzip: should bind a method to Fields object 4`] = `1`;

exports[`Parameters gzip: should bind a method to Fields routingKey  1`] = `"consumerTest"`;

exports[`Parameters gzip: should bind a method to Fields routingKey  2`] = `1`;

exports[`Parameters gzip: should bind a method to Fields routingKey  3`] = `"listenerTest"`;

exports[`Parameters gzip: should bind a method to Fields routingKey  4`] = `1`;

exports[`Parameters gzip: should bind a method to Properties headers  1`] = `
Object {
  "api-auth": "aze&ée(((\\"a&é\\"&é\\"",
}
`;

exports[`Parameters gzip: should bind a method to Properties headers  2`] = `1`;

exports[`Parameters gzip: should bind a method to Properties headers  3`] = `
Object {
  "api-auth": "aze&ée(((\\"a&é\\"&é\\"",
}
`;

exports[`Parameters gzip: should bind a method to Properties headers  4`] = `1`;

exports[`Parameters gzip: should bind a method to Properties object 1`] = `
Object {
  "contentEncoding": "gzip",
  "contentType": "application/json",
  "correlationId": Any<String>,
  "expiration": "5000",
  "headers": Object {},
  "replyTo": Any<String>,
}
`;

exports[`Parameters gzip: should bind a method to Properties object 2`] = `1`;

exports[`Parameters gzip: should bind a method to Properties object 3`] = `
Object {
  "appId": undefined,
  "clusterId": undefined,
  "contentEncoding": "gzip",
  "contentType": "application/json",
  "correlationId": undefined,
  "deliveryMode": undefined,
  "expiration": undefined,
  "headers": Object {},
  "messageId": undefined,
  "priority": undefined,
  "replyTo": undefined,
  "timestamp": undefined,
  "type": undefined,
  "userId": undefined,
}
`;

exports[`Parameters gzip: should bind a method to Properties object 4`] = `1`;

exports[`Parameters: should bind a method parameter to message 1`] = `
Object {
  "content": Any<Object>,
  "fields": Object {
    "consumerTag": Any<String>,
    "deliveryTag": 1,
    "exchange": "",
    "redelivered": false,
    "routingKey": "consumerTest",
  },
  "properties": Object {
    "contentType": "application/json",
    "correlationId": Any<String>,
    "expiration": "5000",
    "headers": Object {},
    "replyTo": Any<String>,
  },
}
`;

exports[`Parameters: should bind a method parameter to message 2`] = `1`;

exports[`Parameters: should bind a method parameter to message 3`] = `
Object {
  "content": Any<Object>,
  "fields": Object {
    "consumerTag": Any<String>,
    "deliveryTag": 3,
    "exchange": "",
    "redelivered": false,
    "routingKey": "listenerTest",
  },
  "properties": Object {
    "appId": undefined,
    "clusterId": undefined,
    "contentEncoding": undefined,
    "contentType": "application/json",
    "correlationId": undefined,
    "deliveryMode": undefined,
    "expiration": undefined,
    "headers": Object {},
    "messageId": undefined,
    "priority": undefined,
    "replyTo": undefined,
    "timestamp": undefined,
    "type": undefined,
    "userId": undefined,
  },
}
`;

exports[`Parameters: should bind a method parameter to message 4`] = `1`;

exports[`Parameters: should bind a method parameter to the content and with object 1`] = `"cool"`;

exports[`Parameters: should bind a method parameter to the content and with object 2`] = `1`;

exports[`Parameters: should bind a method parameter to the content and with object 3`] = `"cool"`;

exports[`Parameters: should bind a method parameter to the content and with object 4`] = `1`;

exports[`Parameters: should bind a method parameter to the content key not exist and call with buffer string 1`] = `undefined`;

exports[`Parameters: should bind a method parameter to the content key not exist and call with buffer string 2`] = `1`;

exports[`Parameters: should bind a method parameter to the content key not exist and call with buffer string 3`] = `undefined`;

exports[`Parameters: should bind a method parameter to the content key not exist and call with buffer string 4`] = `1`;

exports[`Parameters: should bind a method parameter to the content key not exist and with buffer object 1`] = `undefined`;

exports[`Parameters: should bind a method parameter to the content key not exist and with buffer object 2`] = `1`;

exports[`Parameters: should bind a method parameter to the content key not exist and with buffer object 3`] = `undefined`;

exports[`Parameters: should bind a method parameter to the content key not exist and with buffer object 4`] = `1`;

exports[`Parameters: should bind a method parameter to the content key not exist and with object 1`] = `undefined`;

exports[`Parameters: should bind a method parameter to the content key not exist and with object 2`] = `1`;

exports[`Parameters: should bind a method parameter to the content key not exist and with object 3`] = `undefined`;

exports[`Parameters: should bind a method parameter to the content key not exist and with object 4`] = `1`;

exports[`Parameters: should bind a method parameter to the content key not exist and with string 1`] = `undefined`;

exports[`Parameters: should bind a method parameter to the content key not exist and with string 2`] = `1`;

exports[`Parameters: should bind a method parameter to the content key not exist and with string 3`] = `undefined`;

exports[`Parameters: should bind a method parameter to the content key not exist and with string 4`] = `1`;

exports[`Parameters: should bind a method parameter to the content object and call with buffer string 1`] = `"buffertest"`;

exports[`Parameters: should bind a method parameter to the content object and call with buffer string 2`] = `1`;

exports[`Parameters: should bind a method parameter to the content object and call with buffer string 3`] = `"buffertest"`;

exports[`Parameters: should bind a method parameter to the content object and call with buffer string 4`] = `1`;

exports[`Parameters: should bind a method parameter to the content object and with buffer object 1`] = `
Object {
  "test": "buffertest",
}
`;

exports[`Parameters: should bind a method parameter to the content object and with buffer object 2`] = `1`;

exports[`Parameters: should bind a method parameter to the content object and with buffer object 3`] = `
Object {
  "test": "buffertest",
}
`;

exports[`Parameters: should bind a method parameter to the content object and with buffer object 4`] = `1`;

exports[`Parameters: should bind a method parameter to the content object and with object 1`] = `
Object {
  "test": "cool",
}
`;

exports[`Parameters: should bind a method parameter to the content object and with object 2`] = `1`;

exports[`Parameters: should bind a method parameter to the content object and with object 3`] = `
Object {
  "test": "cool",
}
`;

exports[`Parameters: should bind a method parameter to the content object and with object 4`] = `1`;

exports[`Parameters: should bind a method parameter to the content object and with string 1`] = `"test"`;

exports[`Parameters: should bind a method parameter to the content object and with string 2`] = `1`;

exports[`Parameters: should bind a method parameter to the content object and with string 3`] = `"test"`;

exports[`Parameters: should bind a method parameter to the content object and with string 4`] = `1`;

exports[`Parameters: should bind a method to Fields object 1`] = `
Object {
  "consumerTag": Any<String>,
  "deliveryTag": Any<Number>,
  "exchange": Any<String>,
  "redelivered": Any<Boolean>,
  "routingKey": "consumerTest",
}
`;

exports[`Parameters: should bind a method to Fields object 2`] = `1`;

exports[`Parameters: should bind a method to Fields object 3`] = `
Object {
  "consumerTag": Any<String>,
  "deliveryTag": Any<Number>,
  "exchange": Any<String>,
  "redelivered": Any<Boolean>,
  "routingKey": "listenerTest",
}
`;

exports[`Parameters: should bind a method to Fields object 4`] = `1`;

exports[`Parameters: should bind a method to Fields routingKey  1`] = `"consumerTest"`;

exports[`Parameters: should bind a method to Fields routingKey  2`] = `1`;

exports[`Parameters: should bind a method to Fields routingKey  3`] = `"listenerTest"`;

exports[`Parameters: should bind a method to Fields routingKey  4`] = `1`;

exports[`Parameters: should bind a method to Properties headers  1`] = `
Object {
  "api-auth": "aze&ée(((\\"a&é\\"&é\\"",
}
`;

exports[`Parameters: should bind a method to Properties headers  2`] = `1`;

exports[`Parameters: should bind a method to Properties headers  3`] = `
Object {
  "api-auth": "aze&ée(((\\"a&é\\"&é\\"",
}
`;

exports[`Parameters: should bind a method to Properties headers  4`] = `1`;

exports[`Parameters: should bind a method to Properties object 1`] = `
Object {
  "contentType": "application/json",
  "correlationId": Any<String>,
  "expiration": "5000",
  "headers": Object {},
  "replyTo": Any<String>,
}
`;

exports[`Parameters: should bind a method to Properties object 2`] = `1`;

exports[`Parameters: should bind a method to Properties object 3`] = `
Object {
  "appId": undefined,
  "clusterId": undefined,
  "contentEncoding": undefined,
  "contentType": "application/json",
  "correlationId": undefined,
  "deliveryMode": undefined,
  "expiration": undefined,
  "headers": Object {},
  "messageId": undefined,
  "priority": undefined,
  "replyTo": undefined,
  "timestamp": undefined,
  "type": undefined,
  "userId": undefined,
}
`;

exports[`Parameters: should bind a method to Properties object 4`] = `1`;
